src/main/scala/io/iohk/ethereum/forkid/ForkIdValidator.scala

```diff
@@ -0,0 +1,129 @@
+case object Connect extends ForkIdValidationResult
+case object ErrRemoteStale extends ForkIdValidationResult
+case object ErrLocalIncompatibleOrStale extends ForkIdValidationResult
+
+object ForkIdValidator {
+
+  implicit val unsafeLogger = Slf4jLogger.getLogger[Task]
+
+  val maxUInt64 = (BigInt(0x7fffffffffffffffL) << 1) + 1 // scalastyle:ignore magic.number
+
+  def validatePeer[F[_]: Monad: Logger](
+      genesisHash: ByteString,
+      config: BlockchainConfig
+  )(currentHeight: BigInt, remoteForkId: ForkId): F[ForkIdValidationResult] = {
+    val forks = ForkId.gatherForks(config)
+    validatePeer[F](genesisHash, forks)(currentHeight, remoteForkId)
+  }
+
+  def validatePeer[F[_]: Monad: Logger](
+      genesisHash: ByteString,
+      forks: List[BigInt]
+  )(currentHeight: BigInt, remoteId: ForkId): F[ForkIdValidationResult] = {
+    val checksums: Vector[BigInt] = calculateCheckchecksums(genesisHash, forks)
+
+    // find the first unpassed fork and it's index
+    val (unpassedFork, i) = (forks :+ maxUInt64).zipWithIndex.find { case (fork, _) => currentHeight < fork }.get
+
+    // The checks are left biased -> whenever a result is found we need to short circuit
+    val validate: F[Option[ForkIdValidationResult]] = (for {
+      _ <- liftF(Logger[F].trace(s"Before checkMatchingHashes"))
+      matching <- fromEither[F](
+        checkMatchingHashes(checksums, remoteId, currentHeight, i).toLeft("hashes didn't match")
+      )
+      _ <- liftF(Logger[F].trace(s"checkMatchingHashes result: $matching"))
+      _ <- liftF(Logger[F].trace(s"Before checkSubset"))
+      sub <- fromEither[F](checkSubset(checksums, forks, remoteId, i).toLeft("not in subset"))
+      _ <- liftF(Logger[F].trace(s"checkSubset result: $sub"))
+      _ <- liftF(Logger[F].trace(s"Before checkSuperset"))
+      sup <- fromEither[F](checkSuperset(checksums, remoteId, i).toLeft("not in superset"))
+      _ <- liftF(Logger[F].trace(s"checkSuperset result: $sup"))
+      _ <- liftF(Logger[F].trace(s"No check succeeded"))
+      res <- fromEither[F](Either.left[ForkIdValidationResult, Unit](ErrLocalIncompatibleOrStale))
+    } yield (res)).value
+      .map(_.swap)
+      .flatMap(res => Logger[F].debug(s"Validation result is: $res") >> Monad[F].pure(res.toOption))
+
+    Logger[F].debug(s"Validating $remoteId") >>
+      Logger[F].trace(s"Forks list: $forks") >>
+      Logger[F].trace(s"Unpassed fork $unpassedFork was found at index $i") >>
+      validate.map(_.getOrElse(Connect)) // Impossible to say if nodes are compatible, so we need to allow to connect
+  }
+
+  private def calculateCheckchecksums(
+      genesisHash: ByteString,
+      forks: List[BigInt]
+  ): Vector[BigInt] = {
+    val crc = new CRC32()
+    crc.update(genesisHash.asByteBuffer)
+    val genesisChecksum = BigInt(crc.getValue())
+
+    genesisChecksum +: (forks.map { fork =>
+      crc.update(bigIntToBytes(fork, 8))
+      BigInt(crc.getValue())
+    }).toVector
```

Comment:
This code perhaps _looks_ functional but it's really relying on the mutable CRC instance. It's probably more performant this way, but it could be more descriptive using `List.unfold` or similar. (Eg. first produce a collection of collections to produce the CRC with, then flatten it.)

(But this is just a suggestion ofc.)

Reply:
This code isn't functional because of a side effect in `map`. It is because of the nature of Java's `CRC` as you correctly noted. This however doesn't break the referential transparency of `calculateChecksum`, so the function itself remains `pure`.